### PR TITLE
Change how wptrunner computes reftests

### PR DESCRIPTION
**Pull request by gsnedders**
This preserves the current data structures within wptrunner as it
passes them onwards to Marionette currently in the Firefox case, and
as such changing them involves changes within Firefox too.

Those changes within Firefox are ones we should make, and is tracked
by https://bugzilla.mozilla.org/show_bug.cgi?id=1598340, before
changing how reftests are represented within wptrunner.

----

This is the update for https://github.com/web-platform-tests/rfcs/pull/15.

That all said, given we necessarily have to regenerate the manifest for this, we may want to re-evaluate how we store references within the manifest, and whether we want to store a list of pass conditions (like #11345 did with our earlier, much more complex logic).

This PR is also gonna need a bunch of test updates; all of the infrastructure tests for reftest chains etc. now inevitably fail.

### CODE REVIEW

**Comment by gsnedders** [comment]
@jgraham why hasn't TC started running this? :(

**Comment by gsnedders** [comment]
https://community-tc.services.mozilla.com/tasks/groups/FrPCpENuTi-JOgS2Ixyq2w and https://community-tc.services.mozilla.com/tasks/groups/XGmpPteSTGCOEg9U7W8e5g are full Fierfox/Chrome runs for e485b6f47f41310cfedd3d4993a19343d1fd0d9f

File: infrastructure/reftest/reftest_multiple_mismatch-1.html [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<title>AND match that should fail</title>
+<link rel=mismatch href=red.html>
```
Comment (jgraham) [inline infrastructure/reftest/reftest_multiple_mismatch-1.html]: Can we have one with multiple mismatch and multiple match?

**Review by jgraham — COMMENTED** [review]
Generally this looks pretty good. I wonder if there are required docs changes?

**Comment by gsnedders** [comment]
> Generally this looks pretty good. I wonder if there are required docs changes?

Oh. Yes.

https://wpt.fyi/results/css?diff&filter=ADC&run_id=363070004&run_id=349640015 is a comparison of Chrome; https://wpt.fyi/results/?diff&filter=ADC&run_id=353360002&run_id=384230010 is a comparison of Firefox (not quite the right base revision, but almost; the right one doesn't seem to have run against nightly but there are only two SVG tests changed between them).

So some broken tests:

* css/CSS2/selectors/universal-selector-002.xht — matches everything because :first-child matches html:first-child and `color` gets inherited.

Every other regression looks like a correct change with mismatch being AND?

File: tools/wptrunner/wptrunner/wpttest.py [hunk 2]
```diff
@@ -458,38 +451,48 @@ def from_manifest(cls,
...
-                                                      [],
-                                                      None,
-                                                      nodes,
-                                                      references_seen)
-            else:
-                reference = ReftestTest(manifest_file.tests_root,
-                                        ref_url,
+            refs_by_type[ref_type].append(ref_url)
+
+        mismatch_walk = None
```
[the remark this answers is not in the document]
Comment (gsnedders) [reply tools/wptrunner/wptrunner/wpttest.py]: + unit tests

**Comment by stephenmcgruer** [comment]
cc @LukeZielinski as an FYI that we will need to roll wptrunner reasonably soon after this lands. It's a low risk, but it would be possible for a Chromium author to write a now-incorrect reftest downstream with the old wptrunner, which then wouldn't work upstream. Probably not a huge problem, but something to keep an eye on.

**Comment by foolip** [comment]
@LukeZielinski to elaborate on that, we'll "just" need to roll the tools directory and see if it still works. Some small adaptation might be needed in code on our side that interprets the manifest.

**Comment by LukeZielinski** [comment]
@Hexcles as FYI as well. I'm not sure doing another roll of tools in Chromium before the holidays is wise, since a lot of people are disappearing until January starting in the next few days. Which hopefully also means a lot less test authoring happening, but I'd still say we wait until the new year.

**Comment by gsnedders** [comment]
Needs docs changes; otherwise complete.

File: tools/manifest/manifest.py [hunk 3]
```diff
@@ -470,7 +369,28 @@ def from_json(cls, tests_root, obj, types=None):
 
         self._path_hash = {to_os_path(k): v for k, v in iteritems(obj["paths"])}
 
+        # merge reftest_node and reftest
+        # TODO(MANIFESTv7): remove this condition
```
Comment (jgraham) [inline tools/manifest/manifest.py]: Now `MANIFESTv8`.

**Review by jgraham — CHANGES_REQUESTED** [review]

**Review by jgraham — APPROVED** [review]

**Pull request merged**